CONTRIBUTING.md

```diff
@@ -204,6 +205,56 @@ library maintainers on the exclusion from this rule. In such cases there is a
 requirement to test unsafe code with sanitizers including Miri.
 
 
+### Policy
+
+We have various `rust-bitcoin` specific coding styles and conventions that are
+grouped here loosely under the term 'policy'. These are things we try to adhere
+to but that you should not need to worry too much about if you are a new
+contributor. Think of this as a place to collect group knowledge that exists in
+the various PRs over the last few years.
+
+
+#### Errors
+
+Return as much context as possible with errors e.g., if an error was encountered
+parsing a string include the string in the returned error type.
```

Comment:
Error types should be `non_exhaustive` and have private fields unless you're very confident they won't change.
Error types should not commit to implementing traits they may not be able to implement int the future, especially `Copy` but possibly also `Eq` and `Ord`. They should however commit to `Send` and `Sync`.
If the error contains source implement `Display` using `write_err!()` macro.
If a function consumes costly-to-compute input (allocations are also considered costly) it should return it back in the error type.
Errors should have `Error` suffix and conditionally implement `std::error::Error` if they are public.

Reply:
Moving this thread over to the error epic

Reply:
Should we have a small set of derives that are used on all error types? `Clone, Debug, PartialEq`?

Reply:
> They should however commit to `Send` and `Sync`

I don't know what action is required to achieve this, I imagine other readers of the doc may not either. Can you explain further please?

Reply:
`Debug` is the only one I'd guarantee. `Clone` is very likely for most error types but there could be exception when we want to return a resource instead of consuming it. (OTOH `&mut` is probably better)

> I don't know what action is required to achieve this

Not putting things like `Cell` or `Rc` into `Error` types. Sadly it could happen by accident without the compiler complaining and we may break the user code. Maybe we should have something like this:

```rust
// #[allow(unused)] may be required
trait AssertSendSync: Send + Sync {
}

// Perhaps we could also check other traits
macro_rules! check_pub_error {
    ($type:ty) => {
        impl AssertSendSync for $ty {}
    }
}
```

Reply:
Done! https://github.com/rust-bitcoin/rust-bitcoin/pull/1148